azafea/event_processors/endless/metrics/v3/model/__init__.py

```diff
@@ -570,6 +570,64 @@ def _get_fields_from_payload(payload: GLib.Variant) -> Dict[str, Any]:
         }
 
 
+class SplitFlatpakRepoStats(SingularEvent):
+    """Statistics for splitting the Flatpak repo from the OSTree repo.
+
+    Prior to Endless OS 4, both the OS and Flatpaks were stored in a
+    single, shared OSTree repository. On the first boot into Endless OS
+    4 or later, these are split into two repositories, one for the OS
+    and one for Flatpaks. This can take a long time, and blocks the boot
+    process, so we report this event to quantify how long it takes and
+    why.
+
+    This is a one-time operation which only runs if needed, so this
+    event is sent at most once per Endless OS system.
+
+    :UUID name: ``SPLIT_FLATPAK_REPO_STATS`` in eos-boot-helper
+
+    .. versionadded:: 4.0.1
+    """
+    __tablename__ = 'split_flatpak_repo_stats_v3'
+    __event_uuid__ = '880fd091-2e68-4bd2-baa1-f6810fabcc1c'
+    __payload_type__ = '(uuuuuuutut)'
+
+    #: time elapsed in seconds
+    elapsed = Column(Integer, nullable=False)
+    #: number of OS OSTree refs duplicated
+    num_os_refs = Column(Integer, nullable=False)
+    #: number of Flatpak OSTree refs duplicated
+    num_flatpak_refs = Column(Integer, nullable=False)
+    #: number of other OSTree refs duplicated
+    num_other_refs = Column(Integer, nullable=False)
+    #: number of OSTree objects duplicated
+    num_objects = Column(Integer, nullable=False)
+    #: number of OSTree static deltas duplicated
+    num_deltas = Column(Integer, nullable=False)
+    #: number of Flatpak apps duplicated
+    num_apps = Column(Integer, nullable=False)
+    #: size of Flatpak apps duplicated in bytes
+    size_apps = Column(BigInteger, nullable=False)
+    #: number of Flatpak runtimes duplicated
+    num_runtimes = Column(Integer, nullable=False)
+    #: size of Flatpak runtimes duplicated in bytes
+    size_runtimes = Column(BigInteger, nullable=False)
+
+    @staticmethod
+    def _get_fields_from_payload(payload: GLib.Variant) -> Dict[str, Any]:
+        return {
+            'elapsed': payload.get_child_value(0).get_uint32(),
+            'num_os_refs': payload.get_child_value(1).get_uint32(),
+            'num_flatpak_refs': payload.get_child_value(2).get_uint32(),
+            'num_other_refs': payload.get_child_value(3).get_uint32(),
+            'num_objects': payload.get_child_value(4).get_uint32(),
+            'num_deltas': payload.get_child_value(5).get_uint32(),
+            'num_apps': payload.get_child_value(6).get_uint32(),
+            'size_apps': payload.get_child_value(7).get_uint64(),
+            'num_runtimes': payload.get_child_value(8).get_uint32(),
+            'size_runtimes': payload.get_child_value(9).get_uint64(),
```

Comment:
PostgreSQL doesn't support 64-bit unsigned integers. The largest type it supports is 64-bit signed int. https://www.postgresql.org/docs/11/datatype-numeric.html

If you try to insert an out-of-range number, the `INSERT` will just fail:

```
azafea=> create temporary table temp_foo (my_large_number bigint);
CREATE TABLE
azafea=> insert into temp_foo values (18446744073709551615); -- 2 ** 64 - 1
ERROR:  bigint out of range
```

There is a `clamp_to_int64` function to work around this. It logs an error if the value is out of range. But the error message is hard to interpret because there is no context provided about where the number came from. Here's an example from searching the logs in Cloudfront:

```
[ERROR] azafea.event_processors.endless.metrics.v2.utils: Clamped integer larger than MAXINT64: 13752137942398757368
```

So at a minimum you want:

```suggestion
            'size_apps': clamp_to_int64(payload.get_child_value(7).get_uint64()),
            'num_runtimes': payload.get_child_value(8).get_uint32(),
            'size_runtimes': clamp_to_int64(payload.get_child_value(9).get_uint64()),
```

but if you were feeling keen you could add a second string parameter to indicate where the number came from.

Another option would be to make these values signed in the wire protocol. On principle this feels wrong because (a) `flatpak_installed_ref_get_installed_size` returns uint64; (b) a size cannot be negative. But it's an option!

Reply:
Ah, the dreaded exabyte flatpak. Since the issue is here, I think it makes more sense to send the proper data over the wire and handle the issue in azafea.

To fixup `clamp_to_int64`, I'll just get `logging` to print the traceback rather than change all the callers:
```
>>> try:
...     raise ValueError('fail')
... except ValueError:
...     logging.exception('i guess i can let it go')
... 
ERROR:root:i guess i can let it go
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ValueError: fail
```